pkg/commands/git_commands/patch.go

```diff
@@ -254,3 +252,45 @@ func (self *PatchCommands) PullPatchIntoNewCommit(commits []*models.Commit, comm
 	self.PatchManager.Reset()
 	return self.rebase.ContinueRebase()
 }
+
+func restoreOriginalCommit(self *PatchCommands, originalCommitSha string) error {
+	// We first need to "git rm" the files in the patch; this is really only
+	// needed for files that were added in the patch, i.e. that are missing from
+	// the original commit; "git checkout" wouldn't remove these. For files that
+	// are only modified it wouldn't be necessary, but it doesn't hurt either,
+	// so we don't bother making a distinction.
+	for _, filename := range self.PatchManager.AllFilesInPatch() {
+		if _, err := os.Stat(filename); err == nil {
+			if err := self.cmd.New(fmt.Sprintf("git rm %s", self.cmd.Quote(filename))).Run(); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Now checkout the files from the original commit
+	return self.cmd.New(fmt.Sprintf("git checkout %s -- .", originalCommitSha)).Run()
+}
+
+// We have just applied a patch in reverse to discard it from a commit; if we
+// now try to apply the patch again to move it to a later commit, or to the
+// index, then this would conflict "with itself" in case the patch contained
+// only some lines of a range of adjacent added lines. To solve this, we
+// reconstruct a new patch by checking out the original commit again, getting
+// its diff, and resetting it again. This gives us an equivalent patch to the
+// original one, except that it no longer conflicts.
+func restorePatchFromOriginalCommit(self *PatchCommands, commit *models.Commit) (string, error) {
```

Comment:
I've had a go at simplifying this so that no checkout is necessary (by just taking the diff directly) and all the tests pass: https://github.com/jesseduffield/lazygit/commit/b606696e7e18671474ff648a0930815edd46a5e8

Do you know if that commit's approach misses an edge case? If so we ought to add a test for that

Reply:
Nice, thanks. Yes, that's much better, I could have thought of that myself. 😅 

I pushed two fixups for your commit: 8cbd0cfd5781 and 1f0d66546505. I'll now squash your changes down where they belong, give me a minute for that.

I don't think there are any more edge cases here beyond the once I already handle in the existing tests (adding, deleting, and modifying files in a custom patch, that's all I can think of).